Fail unpacking images with xattrs to filesystems without xattr support

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Following up https://github.com/moby/moby/pull/45420#issuecomment-1535047278, I made it a fatal error to untar a file with extended attributes to a destination which does not support extended attributes. I added an escape hatch to the `vfs` graph driver to allow users to opt back into the old, broken behaviour.

**- How I did it**
I made ignoring `EPERM` and `ENOTSUPP` errors from `lsetxattr` when extracting a file from a tarball conditional on a new `BestEffortXattrs` flag in `archive.TarOptions` and plumbed the flag through to the `vfs` driver. I added a new storage option `vfs.xattrs=i_want_broken_containers` to the driver to set the flag. A warning is logged at daemon startup and in `docker info` output if this storage option is enabled.

**- How to verify it**
With the new unit test

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Unpacking layers containing files with extended attributes onto a filesystem which does not support extended attributes will now fail. Previously, the extended attributes would be discarded if they could not be set, which could potentially result in containers malfunctioning. Most storage drivers already require the backing filesystem to support extended attributes so this change should have no impact on most users. The `vfs` driver is a notable exception; the previous behaviour can be restored by configuring it with `--storage-opt vfs.xattrs=i_want_broken_containers`. Enabling this storage option is highly discouraged and bug reports submitted against a daemon with this option set will be summarily closed without comment.

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah reviewed (changes requested):
we discussed this, and because vfs is also a bit of a "last resort", we'd like to provide an escape-hatch for this case ("I_KNOW_WHAT_IM_DOING_GET_ME_OUT_OF_HERE")

thaJeztah commented:
Also for "future  self" (and other visitors); we initially had a concern "what happens if I pull an image, and it's extracted to a filesystem without xattr support (so extended attributes get discarded), and after that I would push that image?
But that's not a concern; when the image layers (tar) are constructed to push, only the file content is read from disk, all other properties are taken from the tar split headers (so even though the extended attributes are missing on the filesystem, the layers will re-apply them before pushing).

corhere committed:
pkg/system: return rich errors from L(g|s)etxattr

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
pkg/archive: bail if setting xattrs is unsupported

Extended attributes are set on files in container images for a reason.
Fail to unpack if extended attributes are present in a layer and setting
the attributes on the unpacked files fails for any reason.

Add an option to the vfs graph driver to opt into the old behaviour
where ENOTSUPP and EPERM errors encountered when setting extended
attributes are ignored. Make it abundantly clear to users and anyone
triaging their bug reports that they are shooting themselves in the
foot by enabling this option.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -98,6 +112,11 @@ func (d *Driver) parseOptions(options []string) error {
 			if err = d.setQuotaOpt(uint64(size)); err != nil {
 				return errdefs.InvalidParameter(errors.Wrap(err, "failed to set option size for vfs"))
 			}
+		case "xattrs":
```
As discussed, this could be `vfs.xattrs`

tianon reviewed (approved):
:eyes: (looks good minus the naming comment Seb just made)

thaJeztah commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -98,6 +112,11 @@ func (d *Driver) parseOptions(options []string) error {
 				return errdefs.InvalidParameter(errors.Wrap(err, "failed to set option size for vfs"))
 			}
+		case "xattrs":
+			if val != "i_want_broken_containers" {
```
We discussed the naming of this, and it looks silly, but it makes it without doubt clear "don't use this... ever", so let's go for it

thaJeztah reviewed (approved):
"LGTM" after changing the `vfs`  prefix, and perhaps a `logrus.Warn` if the setting set.

thaJeztah commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -60,16 +64,26 @@ func Init(home string, options []string, idMap idtools.IdentityMapping) (graphdr
+	if d.bestEffortXattrs {
+		return [][2]string{
+			{"Extended filesystem attributes", "discarded if not supported by the backing filesystem"},
+			{"Tainted configuration", "true"},
+			{"Warranty", "void"},
```
There's no warranty in ASL2, so probably could be dropped 😉

corhere commented:
Note for reviewers: all system-info warnings are logged by the daemon during startup in addition to being included in the `/info` response.
https://github.com/moby/moby/blob/ccd834ea254da112673968a620386717f38bf31a/daemon/daemon.go#L1136-L1139

neersighted reviewed (approved):
LGTM, and the warning language seems to strike a good balance now.

thaJeztah commented:
Test failure looks related

```
=== FAIL: daemon/graphdriver/vfs TestXattrUnsupportedByBackingFS/xattrs=i_want_broken_containers (0.00s)
    vfs_test.go:98: assertion failed: error is not nil: unknown option xattrs for vfs

neersighted commented:
...and that was the wrong button (I was trying to click 'Show all checks' but my brain wasn't quite there yet) -- @corhere, could you figure out the test issue in a follow-up?
